### PR TITLE
More Advanced Logging & Error Handling

### PR DESCRIPTION
**Pull request by DCMattyG**
This update adds more advanced logging & error handling capabilities to both the Engine and the Scaler. The Scaler modules now bubble up their errors to the Scaler function, which in turn properly forms and throws a PowerShell error. Added ERRORDATA output which can be captured by Application Insights for monitoring/alerting. Adjusted the Scaler to only dequeue a message once, so a failure won't trigger the Scaler to run 4 additional times. Finally, I cleaned up a few minors spelling/grammatical items, as well as set some nice-to-have defaults in the ARM Template.

### CODE REVIEW

File: functions/engine/BellhopEngine.cs [hunk 1]
```diff
@@ -139,24 +139,33 @@ public static void Validate()
...
         public string ResourceGroup { get; }
-        public string Subscription { get; }
+        public string SubscriptionId { get; }
+        public string SubscriptionName { get; }
         public string CurrentState { get; }
         public string TargetState { get; }
         public string ResizeAction { get; }
         public JObject GraphData { get; }
-        public List<KeyValuePair<object, string>> _errors = new List<KeyValuePair<object, string>>();
+        public List<Exception> _errors = new List<Exception>();
```
Comment (NillsF) [inline functions/engine/BellhopEngine.cs:159]: Why are you using an `_` in front of errors? Based on what I'm reading online, `_` would be used for private variables.

File: templates/azuredeploy.json [hunk 2]
```diff
@@ -10,12 +10,14 @@
         },
         "location": {
             "type": "string",
+            "defaultValue": "[deployment().location]",
```
Comment (NillsF) [inline templates/azuredeploy.json:13]: Nice!

**Review by NillsF — COMMENTED** [review]
Haven't tested yet, so not hitting approve yet.

One question that came up as I'm reviewing this:
We have another parameter section in the nested template. Do we actually want to pass parameters to the nested template, or just re-use the parameters from the subscription-level template (default behaviour as per [this document](https://docs.microsoft.com/en-us/azure/azure-resource-manager/templates/linked-templates?tabs=azure-powershell#expression-evaluation-scope-in-nested-templates))? Do they ever change between the subscription template and the RG template?

File: functions/engine/BellhopEngine.cs (hunk 1 above)
Comment (DCMattyG) [reply functions/engine/BellhopEngine.cs:159]: You are correct, this should be changed to a Private variable. Good catch.

File: templates/azuredeploy.json (hunk 2 above)
Comment (DCMattyG) [reply templates/azuredeploy.json:13]: Word!

File: functions/engine/BellhopEngine.cs (hunk 1 above)
Comment (DCMattyG) [reply functions/engine/BellhopEngine.cs]: Just pushed an update to switch this to a private variable instead of public.

**Comment by DCMattyG** [comment]
Hey @NillsF, I was originally using the "outer" scope, but that caused some issues in how some of the dependencies worked, and the scoping of certain resources. In the future, I foresee breaking this template up into multiple smaller Linked templates to make it easier to update. For now, I believe it's better to leave the scope to "inner", but I'm happy to revisit this later and attempt using the "outer" scope once again.

File: docs/scaler/modules/appservice.md [hunk 3]
```diff
@@ -59,11 +59,11 @@ An example of the message sent to the queue by the engine function
             "subscription": "<SUBSCRIPTION-ID>",
             "kind": "linux",
             "tags": {
-                "resize-Enable": "True"
+                "resize-Enable": "True",
                 "resize-StartTime": "Friday 7PM",
                 "resize-EndTime": "Monday 6AM",
                 "setState-WorkerSize": "Small",
-                "setState-Tier": "Basic",
+                "setState-Tier": "Basic"
```
Comment (CloudViking) [inline docs/scaler/modules/appservice.md:66]: Nice catch on these missing or extra commas in the docs... It's those little details! 👍

**Review by CloudViking — APPROVED** [review]
This is big for the project, the logging is clean. Nice work!

File: docs/scaler/modules/appservice.md (hunk 3 above)
Comment (DCMattyG) [reply docs/scaler/modules/appservice.md:66]: Thanks man!

**Review by NillsF — CHANGES_REQUESTED** [review]
Discovered a bug: when set-state tag is named incorrectly with ASP, the scaling operation still happens (i.e. no error). Please add a check in the ASP scaler to verify that if setstate is empty it should throw an error.

**Comment by DCMattyG** [comment]
Hey @NillsF, really good catch! I added some additional validations to the Scaler to make sure that setData/saveData are not Null or Empty. Much appreciated sir.

**Review by NillsF — APPROVED** [review]
Retested after your latest changes. LGTM

**Pull request merged**